orderer/common/follower/follower_chain.go

```diff
@@ -426,6 +426,129 @@ func (c *Chain) pullUpToJoin() error {
+func (c *Chain) pullUntilNextConfig() (*common.Block, error) {
+	// This puller is built from the tip of the ledger. When we get a new config block we exit the method. That will
+	// destroy the puller. Next time we enter, a new puller is built, taking in the new config.
+	puller, err := c.createBlockPullerFunc()
+	if err != nil {
+		c.logger.Errorf("Error creating block puller: %s", err)
+		return nil, errors.Wrap(err, "error creating block puller")
+	}
+	defer puller.Close()
+
+	var configBlock *common.Block
+	for configBlock == nil {
+		endpoint, latestHeight, err := cluster.LatestHeightAndEndpoint(puller)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get latest height and endpoint")
+		}
+		c.logger.Debugf("Orderer endpoint %s has the biggest ledger height: %d", endpoint, latestHeight)
+
+		if latestHeight <= c.ledgerResources.Height() {
+			c.logger.Debugf("My height: %d, latest height: %d; going to wait %v for latest height to grow",
+				c.ledgerResources.Height(), latestHeight, c.retryInterval)
+			select {
+			case <-c.stopChan:
+				return nil, errors.New("stopped while waiting for latest height to grow")
+			case <-c.timeAfter(c.retryInterval):
+				c.increaseRetryInterval()
+				continue
+			}
+		}
+
+		configBlock, err = c.pullUntilLatestOrConfig(puller, latestHeight)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to pull until latest height or config")
+		}
+		c.resetRetryInterval()
+	}
+
+	return configBlock, nil
+}
+
+func (c *Chain) pullUntilLatestOrConfig(puller ChannelPuller, latestHeight uint64) (*common.Block, error) {
+	firstBlockToPull := c.ledgerResources.Height()
+	var nextBlock = c.ledgerResources.Block(firstBlockToPull - 1)
+	if nextBlock == nil {
+		return nil, errors.Errorf("cannot retrieve previous block %d", firstBlockToPull-1)
+	}
+	var actualPrevHash = protoutil.BlockHeaderHash(nextBlock.Header)
+
+	// Pull until the latest height or a config block
+	for seq := firstBlockToPull; seq < latestHeight; seq++ {
+		select {
+		case <-c.stopChan:
+			c.logger.Warnf("Stopped while pulling blocks: from %d until %d, last pulled %d",
+				firstBlockToPull, latestHeight, nextBlock.Header.Number)
+			return nil, errors.New("stopped while pulling blocks")
+		default:
+			nextBlock = puller.PullBlock(seq)
+			if nextBlock == nil {
+				return nil, errors.Wrapf(cluster.ErrRetryCountExhausted, "failed to pull block %d", seq)
+			}
+			reportedPrevHash := nextBlock.Header.PreviousHash
+			if !bytes.Equal(reportedPrevHash, actualPrevHash) {
+				return nil, errors.Errorf("block header mismatch on sequence %d, expected %x, got %x",
+					nextBlock.Header.Number, actualPrevHash, reportedPrevHash)
+			}
+			actualPrevHash = protoutil.BlockHeaderHash(nextBlock.Header)
+			if err := c.ledgerResources.Append(nextBlock); err != nil {
+				return nil, errors.Wrapf(err, "failed to append block %d to the ledger", nextBlock.Header.Number)
+			}
+			c.resetRetryInterval()
```

Comment:
It seems odd to sprinkle timer resets throughout code which does not otherwise interact with the timer.  Would the caller not be a better place to do this?

Reply:
The resets are not "sprinkled"... they are placed after a successful pull, just after Append.
The retryInterval is increased on successive failed attempts in the outer method and reset on the first successful pull. Successful pulls remain in this method (loop), so there is no chance for the outer one to reset in the right time.

Reply:
I think part of the issue I'm having with this function is that it acts both by return value, and by side effect. 

You can see up on line 459:

```
		￼configBlock, err = c.pullUntilLatestOrConfig(puller, latestHeight)
￼		if err != nil {
￼			return nil, errors.Wrap(err, "failed to pull until latest height or config")
￼		}
￼		c.resetRetryInterval()
```

Which, implies to me that `pullUntilLatestOrConfig` returns a config block, or nil, or an error.  And, based on the success or failure of this function, we reset the timer, or not.

But, it's actually not possible to return without error without already having reset the timer in the function itself.  So, why are we resetting it again on the outside?

Similarly, inside, we are setting the member `c.lastConfig`, and then we are also returning what we just set as the first return value of this function.

I suspect it's an artifact of starting from the consenter interface, but the follower to me seems like it should have very simple control flow, but it's very hard for me to get it from this code.

1. Pull blocks, until the latest known height.  On error, try again with some backoff logic.
2. Check to see if we're in the consenter set, and if we are, exit.
3. Recheck the block height periodically, until it advances beyond where we are, go to 1.

It seems to me that there's a lot of other behavior here.  We've got a retry timer that's used in the main run loop, which is manipulated throughout the functions it calls into, but not in ways that are necessarily clear.  If an error is returned, is the timer reset? Maybe.  The selected block pulling function is done by function pointer, the timeouts are adjusted based on the pulling type.  To me, this feels like it should be considerably less and much more clear code.

Reply:
Addressed in the last commit.